### PR TITLE
Switch to session locks for maintenance tasks

### PR DESCRIPTION
**Pull request by cevian**

### CODE REVIEW

File: pkg/migrations/sql/idempotent/base.sql [hunk 1]
```diff
@@ -72,26 +72,34 @@ CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.lock_metric_for_maintenance(metric_id
...
+        SELECT pg_try_advisory_lock(12378, metric_id) INTO STRICT res;
 
-        RETURN FOUND;
+        RETURN res;
     ELSE
-        PERFORM m.*
-        FROM SCHEMA_CATALOG.metric m
-        WHERE m.id = metric_id
-        FOR NO KEY UPDATE;
+        PERFORM pg_advisory_lock(12378, metric_id);
```
Comment (davidkohn88) [inline pkg/migrations/sql/idempotent/base.sql:82]: Having a note basically everywhere we use this 12378 thing is probably a good idea, it strikes me as odd every time I see it.

**Review by davidkohn88 — APPROVED** [review]
LGTM. Just the one note on the random constant we're using.

File: pkg/migrations/sql/idempotent/base.sql (hunk 1 above)
Comment (Harkishen-Singh) [reply pkg/migrations/sql/idempotent/base.sql:82]: I think this should be done like how we did below

https://github.com/timescale/promscale/blob/9c6e713a96cdb93154cf319702c52619913e8bd6/pkg/pgmodel/migrate.go#L372

File: pkg/migrations/sql/idempotent/base.sql [hunk 2]
```diff
@@ -72,26 +72,34 @@ CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.lock_metric_for_maintenance(metric_id
...
 $func$
 LANGUAGE PLPGSQL VOLATILE;
 GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.lock_metric_for_maintenance(int, boolean) TO prom_writer;
+
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.unlock_metric_for_maintenance(metric_id int)
+    RETURNS VOID
+AS $func$
+DECLARE
+BEGIN
+    PERFORM pg_advisory_unlock(12378, metric_id);
```
Comment (Harkishen-Singh) [inline pkg/migrations/sql/idempotent/base.sql:96]: this needs to be replaced in a similar way

File: pkg/migrations/sql/idempotent/base.sql [hunk 3]
```diff
@@ -72,26 +72,34 @@ CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.lock_metric_for_maintenance(metric_id
...
 AS $func$
 DECLARE
+    res BOOLEAN;
 BEGIN
     IF NOT wait THEN
-        PERFORM m.*
-        FROM SCHEMA_CATALOG.metric m
-        WHERE m.id = metric_id
-        FOR NO KEY UPDATE SKIP LOCKED;
+        SELECT pg_try_advisory_lock(12378, metric_id) INTO STRICT res;
```
Comment (Harkishen-Singh) [inline pkg/migrations/sql/idempotent/base.sql:78]: this too

**Review by Harkishen-Singh — APPROVED** [review]
Small nit and then we are good to merge.

**Pull request merged**